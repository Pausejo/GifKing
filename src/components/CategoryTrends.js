import React from 'react'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faSync, faArrowTrendUp } from '@fortawesome/free-solid-svg-icons'
import { useTrends } from '../hooks/useTrends';

export const CategoryTrends = ({setCategories}) => {

    const {data: trends, loading} = useTrends();

    const trendClick = (trend) => {
        setCategories(cat => {
            return (!cat.find(c => c === trend.target.textContent)) ? 
                [trend.target.textContent, ...cat] : 
                cat
        });
    }

    return (<>
                <h3 className='trending'><FontAwesomeIcon icon={faArrowTrendUp} /> Trending</h3>
                
                <ol className="filters">
                    {
                        trends.map(t =>
                            <li key={t}>
                                <label onClick={trendClick}>{t}</label>
                            </li>)
                    }
                </ol>

                {loading && <FontAwesomeIcon icon={faSync} spin size='1x' />}
            </>
    )
}
