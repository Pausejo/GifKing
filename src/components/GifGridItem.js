import React from 'react';
import PropTypes from 'prop-types';

export const GifGridItem = ({url, title}) => {

  const handleClick = () => {    
    navigator.clipboard.writeText(url);
  }

  return <div className="card" onClick={handleClick}>
            <h2 className="title">{title}</h2>
            <img src={url} alt={title} />
        </div>
}

GifGridItem.propTypes = {
  url: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
}