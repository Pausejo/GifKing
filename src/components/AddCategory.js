import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { CategoryTrends } from './CategoryTrends';

export const AddCategory = ({setCategories}) => {

    const [inputValue, setInputValue] = useState('');

    const handleInputChange = (e)=>{
        setInputValue(e.target.value);        
    }

    const handleSubmit = (e)=>{
        e.preventDefault();

        if (inputValue.length>2) {
            setCategories(cat => {
                return (!cat.find(c => c === inputValue)) ? 
                    [inputValue, ...cat] : 
                    cat
            });
            setInputValue('');
        }
    }

    return (<>
            <form data-testid="form" onSubmit={handleSubmit}>
                <div className="row">
                    <input
                        className="balloon"
                        id="category"
                        type="text"
                        value={inputValue}
                        onChange={handleInputChange}
                        placeholder="Ingrese la categorÃ­a" />
                        <label htmlFor="category">Category</label>
                </div>
            </form>
            <CategoryTrends setCategories={setCategories}/>
        </>
    );
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired,
}