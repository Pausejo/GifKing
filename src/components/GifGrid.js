import React from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faCircleMinus, faSync } from '@fortawesome/free-solid-svg-icons';
import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifGridItem } from './GifGridItem';
import PropTypes from 'prop-types';

export const GifGrid = ({category, setCategories}) => {
    const {data: images, loading} = useFetchGifs(category);

    const deleteCatagory = () => {        
        setCategories((c) => c.filter(i => i!== category));
    }

    return (
         <div role="GifGrid">
            <h2><label onClick={deleteCatagory}><FontAwesomeIcon className="delete" icon={faCircleMinus} /> {category}</label></h2>        

            {loading && <FontAwesomeIcon data-testid="loading" icon={faSync} spin size='4x' />}

            <div className="container">
                { images.map(img => <GifGridItem key={img.id} {...img} />) }            
            </div>
        </div>
    )
}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired,
    setCategories: PropTypes.func.isRequired,
}