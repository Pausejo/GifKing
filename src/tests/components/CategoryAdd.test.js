import { render, screen, fireEvent } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { AddCategory } from '../../components/AddCategory';

describe('Testing CategoryAdd', () => { 

    const setCategories = jest.fn();

    beforeEach(() => {
      jest.clearAllMocks();
    });

    test('should show component correctly', () => { 
        const component = renderer.create(<AddCategory setCategories={setCategories}/>);

        const tree = component.toJSON();

        expect(tree).toMatchSnapshot(); 
     });

     test('should change input', () => { 
        const value = 'Hello World';
        
        render(<AddCategory setCategories={setCategories}/>);

        const input = screen.getByLabelText('Category');

        fireEvent.change(input, {target: {value}})

        fireEvent.keyPress(input,)

        expect(input.value).toBe(value);
        
     });

     test('should not submit', () => {         
        render(<AddCategory setCategories={setCategories}/>);

        const form = screen.getByTestId('form');

        fireEvent.submit(form);

        expect(setCategories).not.toBeCalled();
     });

     test('should submit', () => {         
        render(<AddCategory setCategories={setCategories}/>);

        const value = 'Pepe Viyuela';

        const input = screen.getByLabelText('Category');
        
        fireEvent.change(input, {target: {value}})
        
        const form = screen.getByTestId('form');

        fireEvent.submit(form);

        expect(setCategories).toHaveBeenCalledTimes(1)
        expect(setCategories).toHaveBeenCalledWith(expect.any(Function));

        expect(input.value).toBe('');
     });

 });