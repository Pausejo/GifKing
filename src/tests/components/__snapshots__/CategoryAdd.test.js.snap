// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing CategoryAdd should show component correctly 1`] = `
Array [
  <form
    data-testid="form"
    onSubmit={[Function]}
  >
    <div
      className="row"
    >
      <input
        className="balloon"
        id="category"
        onChange={[Function]}
        placeholder="Ingrese la categorÃ­a"
        type="text"
        value=""
      />
      <label
        htmlFor="category"
      >
        Category
      </label>
    </div>
  </form>,
  <h3
    className="trending"
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-arrow-trend-up "
      data-icon="arrow-trend-up"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 576 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M384 160C366.3 160 352 145.7 352 128C352 110.3 366.3 96 384 96H544C561.7 96 576 110.3 576 128V288C576 305.7 561.7 320 544 320C526.3 320 512 305.7 512 288V205.3L342.6 374.6C330.1 387.1 309.9 387.1 297.4 374.6L191.1 269.3L54.63 406.6C42.13 419.1 21.87 419.1 9.372 406.6C-3.124 394.1-3.124 373.9 9.372 361.4L169.4 201.4C181.9 188.9 202.1 188.9 214.6 201.4L320 306.7L466.7 159.1L384 160z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
     Trending
  </h3>,
  <ol
    className="filters"
  />,
  <svg
    aria-hidden="true"
    className="svg-inline--fa fa-arrows-rotate fa-spin fa-1x "
    data-icon="arrows-rotate"
    data-prefix="fas"
    focusable="false"
    role="img"
    style={Object {}}
    viewBox="0 0 512 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M464 16c-17.67 0-32 14.31-32 32v74.09C392.1 66.52 327.4 32 256 32C161.5 32 78.59 92.34 49.58 182.2c-5.438 16.81 3.797 34.88 20.61 40.28c16.89 5.5 34.88-3.812 40.3-20.59C130.9 138.5 189.4 96 256 96c50.5 0 96.26 24.55 124.4 64H336c-17.67 0-32 14.31-32 32s14.33 32 32 32h128c17.67 0 32-14.31 32-32V48C496 30.31 481.7 16 464 16zM441.8 289.6c-16.92-5.438-34.88 3.812-40.3 20.59C381.1 373.5 322.6 416 256 416c-50.5 0-96.25-24.55-124.4-64H176c17.67 0 32-14.31 32-32s-14.33-32-32-32h-128c-17.67 0-32 14.31-32 32v144c0 17.69 14.33 32 32 32s32-14.31 32-32v-74.09C119.9 445.5 184.6 480 255.1 480c94.45 0 177.4-60.34 206.4-150.2C467.9 313 458.6 294.1 441.8 289.6z"
      fill="currentColor"
      style={Object {}}
    />
  </svg>,
]
`;
